android/src/main/java/com/oney/WebRTCModule/WebRTCDevicesManager.java

```diff
@@ -0,0 +1,247 @@
+
+        private String kind;
+
+        DeviceKind(String kind){
+            this.kind = kind;
+        }
+
+        public String getKind() {
+            return this.kind;
+        }
+    }
+
+    public enum AudioRoute {
+        // Those constants are defined on the webrtc specification
+        // https://w3c.github.io/mediacapture-main/#dom-mediadevicekind
+        ROUTE_BUILT_IN(1),
+        ROUTE_SPEAKER(2),
+        ROUTE_BLUETOOTH(3);
+
+        private int value;
+
+        AudioRoute(int value){
+            this.value = value;
+        }
+
+        public int getValue() {
+            return this.value;
+        }
+
+        static AudioRoute get(int value){
+            return Arrays.stream(AudioRoute.values()).filter(audioRoute -> audioRoute.value == value).findFirst().get();
+        }
+    }
+
+    private enum DeviceType {
+        BLUETOOTH(AudioRoute.ROUTE_BLUETOOTH),
+        HEADSET(AudioRoute.ROUTE_BUILT_IN),
+        SPEAKER(AudioRoute.ROUTE_SPEAKER),
+        EARPIECE(AudioRoute.ROUTE_BUILT_IN),
+        BUILT_IN_MICROPHONE(AudioRoute.ROUTE_BUILT_IN),
+        BUILT_IN_MICROPHONE_SPEAKER(AudioRoute.ROUTE_SPEAKER),
+        CAMERA_USER(null),
+        CAMERA_ENVIRONMENT(null);
+
+        private AudioRoute audioRoute;
+
+        DeviceType(AudioRoute audioRoute){
+            this.audioRoute = audioRoute;
+        }
+    }
+
+    private final CameraEnumerator cameraEnumerator;
+    private AudioManager audioManager;
+    private ReactApplicationContext reactContext;
+
+    public WebRTCDevicesManager(ReactApplicationContext reactContext) {
+        this.reactContext = reactContext;
+        this.audioManager = (AudioManager) reactContext.getSystemService(Context.AUDIO_SERVICE);
+        this.cameraEnumerator = this.createCameraEnumerator();
+    }
+
+    private CameraEnumerator createCameraEnumerator() {
+        boolean camera2supported = false;
+        try {
+            camera2supported = Camera2Enumerator.isSupported(this.reactContext);
+        } catch (Throwable tr) {
+            // Some devices will crash here with: Fatal Exception: java.lang.AssertionError: Supported FPS ranges cannot be null.
+            // Make sure we don't.
+            Log.w(TAG, "Error checking for Camera2 API support.", tr);
+        }
+        CameraEnumerator cameraEnumerator = null;
+        if (camera2supported) {
+            Log.d(TAG, "Creating video capturer using Camera2 API.");
+            cameraEnumerator = new Camera2Enumerator(this.reactContext);
+        } else {
+            Log.d(TAG, "Creating video capturer using Camera1 API.");
+            cameraEnumerator = new Camera1Enumerator(false);
+        }
+        return cameraEnumerator;
+    }
+
+    ReadableArray enumerateDevices() {
+        WritableArray devicesArray = Arguments.createArray();
+        this.fillVideoInputDevices(devicesArray);
+        this.fillAudioInputDevices(devicesArray);
+        this.fillAudioOutputDevices(devicesArray);
```

Comment:
philosophical question: why list output devices in addition to input devices if there's no way for a customer to specify output device? would it be bad if we just listed routes once (and called them inputs for now)?

Reply:
Refactor, now we only have "audio"